test_conformance/subgroups/subhelpers.h

```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+            case ArithmeticOp::mul_: return (cl_double)1;
+            default: log_error("Unknown operation request"); break;
+        }
+        return 0;
+    }
+    static const bool type_supported(cl_device_id device)
+    {
+        return double_ok(device);
+    }
+};
+template <>
+struct TypeManager<cl_double2> : public CommonTypeManager<cl_double2>
+{
+    static const char *name() { return "double2"; }
+    static const char *add_typedef() { return "typedef double2 Type;\n"; }
+    using scalar_type = cl_double;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return double_ok(device);
+    }
+};
+template <>
+struct TypeManager<subgroups::cl_double3>
+    : public CommonTypeManager<subgroups::cl_double3>
+{
+    static const char *name() { return "double3"; }
+    static const char *add_typedef() { return "typedef double3 Type;\n"; }
+    typedef std::true_type is_sb_vector_size3;
+    using scalar_type = cl_double;
+    static const bool type_supported(cl_device_id device)
+    {
+        return double_ok(device);
+    }
+};
+template <>
+struct TypeManager<cl_double4> : public CommonTypeManager<cl_double4>
+{
+    static const char *name() { return "double4"; }
+    static const char *add_typedef() { return "typedef double4 Type;\n"; }
+    using scalar_type = cl_double;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return double_ok(device);
+    }
+};
+template <>
+struct TypeManager<cl_double8> : public CommonTypeManager<cl_double8>
+{
+    static const char *name() { return "double8"; }
+    static const char *add_typedef() { return "typedef double8 Type;\n"; }
+    using scalar_type = cl_double;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return double_ok(device);
+    }
+};
+template <>
+struct TypeManager<cl_double16> : public CommonTypeManager<cl_double16>
+{
+    static const char *name() { return "double16"; }
+    static const char *add_typedef() { return "typedef double16 Type;\n"; }
+    using scalar_type = cl_double;
+    typedef std::true_type is_vector_type;
+    static const bool type_supported(cl_device_id device)
+    {
+        return double_ok(device);
+    }
 };
 
-template <> struct TypeIdentity<double, 1>
+// cl_half
+template <>
+struct TypeManager<subgroups::cl_half>
+    : public CommonTypeManager<subgroups::cl_half>
 {
-    static double val() { return -std::numeric_limits<double>::infinity(); }
+    static const char *name() { return "half"; }
+    static const char *add_typedef() { return "typedef half Type;\n"; }
+    typedef std::true_type is_sb_scalar_type;
+    static subgroups::cl_half identify_limits(ArithmeticOp operation)
+    {
+        switch (operation)
+        {
+            case ArithmeticOp::add_: return { 0x0000 };
+            case ArithmeticOp::max_: return { 0xfc00 };
```

Comment:
Comments on these values would be helpful to translate each into reader-friendly form